server/etcdserver/api/etcdhttp/readiness.go

```diff
@@ -0,0 +1,202 @@
+				failedChecks = append(failedChecks, checkName)
+			} else {
+				fmt.Fprintf(&individualCheckOutput, "[+]%s ok\n", checkName)
+			}
+		}
+		if len(excluded) > 0 {
+			fmt.Fprintf(&individualCheckOutput, "warn: some health checks cannot be excluded: no matches for %s\n", formatQuoted(excluded.List()...))
+		}
+		if len(included) > 0 {
+			fmt.Fprintf(&individualCheckOutput, "warn: some health checks cannot be included: no matches for %s\n", formatQuoted(included.List()...))
+		}
+		// always be verbose on failure
+		if len(failedChecks) > 0 {
+			http.Error(w, fmt.Sprintf("%s%s check failed", individualCheckOutput.String(), reg.path), http.StatusInternalServerError)
+			lg.Error(fmt.Sprintf("%s%s check failed", individualCheckOutput.String(), reg.path))
+			return
+		}
+		lg.Info(fmt.Sprintf("%s check passed:\n%s", reg.path, individualCheckOutput.String()))
+
+		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+		w.Header().Set("X-Content-Type-Options", "nosniff")
+		if _, found := r.URL.Query()["verbose"]; !found {
+			fmt.Fprint(w, "ok")
+			return
+		}
+		individualCheckOutput.WriteTo(w)
+		fmt.Fprintf(w, "%s check passed\n", reg.path)
+	}
+}
+
+// adaptCheckToHandler returns an http.HandlerFunc that serves the provided checks.
+func adaptCheckToHandler(c HealthCheck) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		err := c(r.Context())
+		if err != nil {
+			http.Error(w, fmt.Sprintf("internal server error: %v", err), http.StatusInternalServerError)
+		} else {
+			fmt.Fprint(w, "ok")
+		}
+	})
+}
+
+// formatQuoted returns a formatted string of the health check names,
+// preserving the order passed in.
+func formatQuoted(names ...string) string {
+	quoted := make([]string, 0, len(names))
+	for _, name := range names {
+		quoted = append(quoted, fmt.Sprintf("%q", name))
+	}
+	return strings.Join(quoted, ",")
+}
+
+type stringSet map[string]struct{}
+
+func (s stringSet) List() []string {
+	keys := make([]string, len(s))
+
+	i := 0
+	for k := range s {
+		keys[i] = k
+		i++
+	}
+	return keys
+}
+
+func listToStringSet(list []string) stringSet {
+	set := make(map[string]struct{})
+	for _, s := range list {
+		if len(s) == 0 {
+			continue
+		}
+		set[s] = struct{}{}
+	}
+	return set
+}
+
+// checkAlarm checks if a specific alarm type is active in the server.
+func checkAlarm(srv ServerHealth, at etcdserverpb.AlarmType) func(context.Context) error {
+	return func(ctx context.Context) error {
+		as := srv.Alarms()
+		for _, v := range as {
+			if v.Alarm == at {
+				return fmt.Errorf("Alarm active:%s", at.String())
+			}
+		}
+		return nil
+	}
+}
+
+func checkSerializableRead(srv ServerHealth) func(ctx context.Context) error {
+	return func(ctx context.Context) error {
+		_, err := srv.Range(ctx, &etcdserverpb.RangeRequest{KeysOnly: true, Limit: 1, Serializable: true})
+		if err != nil && err != auth.ErrUserEmpty && err != auth.ErrPermissionDenied {
```

Comment:
I think this check should work even if authorization is enabled.

Reply:
> Do you mean there is a way to override the permission or livez should fail if the requester is not authorized?

@siyuanfoundation I guess what @serathius meant is if authorization is enabled, the serializable check won't exercise the code path it wants to (read a key from mvcc) and return permission error early. 

And I have a idea to fix this existing implementation limitation. 

```
// WithRoot generates and installs a token that can be used as a root credential
```

`ctx = srv.AuthStore().WithRoot(ctx)`

It should generate a token valid for the range request to pass permission check. It has been used in [revokeExpiredLeases](https://github.com/etcd-io/etcd/blob/628b45c09921b11a4f61a9c5dd7b371e76cd5b59/server/etcdserver/server.go#L874). Ideally I want it to be read only root permissions to reduce security risks exposure since the livez check can be triggered from outside. 

However, I suggest we can keep the PR as it is and raise a separate issue to explain it and fix this.